core/cache/watcher.go

```diff
@@ -146,6 +148,188 @@ func (w *ConfigWatcher) configChanged(topic string, value interface{}) {
+}
+
+// applicationCharmURLChange sends a notification if what is saved for its
+// charm lxdprofile changes.  No notification is sent if the pointer begins
+// and ends as nil.
+func (w *MachineAppLXDProfileWatcher) applicationCharmURLChange(topic string, value interface{}) {
+	w.notifyWatcherBase.mu.Lock()
+	var notify bool
+	defer func(notify *bool) {
+		w.notifyWatcherBase.mu.Unlock()
+		if *notify {
+			w.notify()
+		}
+	}(&notify)
+
+	values, ok := value.([]string)
+	if !ok {
+		logger.Errorf("programming error, value not of type []string")
+		return
+	}
+	if len(values) != 2 {
+		logger.Errorf("programming error, 2 values not provided")
+		return
+	}
+	appName, chURL := values[0], values[1]
+	info, ok := w.applications[appName]
+	if ok {
+		ch, err := w.getCharm(chURL)
+		if err != nil {
+
+		}
+		// notify if:
+		// 1. the prior charm had a profile and the new one does not.
+		// 2. the new profile is not empty.
+		if (info.charmProfile != nil && ch.details.LXDProfile.Empty()) ||
+			!ch.details.LXDProfile.Empty() {
+			logger.Tracef("notifying due to change of charm lxd profile for %s, machine-%s", appName, w.machineId)
+			notify = true
+		} else {
+			logger.Tracef("no notification of charm lxd profile needed for %s, machine-%s", appName, w.machineId)
+		}
+		if ch.details.LXDProfile.Empty() {
+			info.charmProfile = nil
+		} else {
+			info.charmProfile = &ch.details.LXDProfile
+		}
+		info.charmURL = chURL
+		w.applications[appName] = info
+	} else {
+		logger.Errorf("not watching %s on machine-%s", appName, w.machineId)
+	}
+}
+
+// unitChange modifies the map of applications being watched when a unit is
+// added or removed from the machine.  Notification is sent if:
+//     1. A new unit whose charm has an lxd profile is added.
+//     2. A unit being removed has a profile and other units
+//        exist on the machine.
+func (w *MachineAppLXDProfileWatcher) unitChange(topic string, value interface{}) {
+	w.notifyWatcherBase.mu.Lock()
+	var notify bool
+	defer func(notify *bool) {
+		w.notifyWatcherBase.mu.Unlock()
+		if *notify {
+			logger.Tracef("notifying due to add/remove unit requires lxd profile change")
+			w.notify()
+		}
+	}(&notify)
+
+	names, okString := value.([]string)
+	unit, okUnit := value.(*Unit)
+	switch {
+	case okString:
+		logger.Tracef("Stop watching %q", names)
+		notify = w.removeUnit(names)
+	case okUnit:
+		if w.machineId != unit.details.MachineId {
+			logger.Tracef("not the machine being watched")
+			return
+		}
+		logger.Tracef("Start watching %q", unit.details.Name)
+		notify = w.addUnit(unit)
+	default:
+		logger.Errorf("programming error, value not of type *Unit or []string")
+	}
+}
+
+func (w *MachineAppLXDProfileWatcher) addUnit(unit *Unit) bool {
+	_, ok := w.applications[unit.details.Application]
+	if !ok {
+		info := appInfo{
+			charmURL: unit.details.CharmURL,
+			units:    set.NewStrings(unit.details.Name),
+		}
+		ch, err := w.getCharm(unit.details.CharmURL)
+		if err != nil {
+
```

Comment:
Missing handling error?

Reply:
will be resolved in pr 9966 lxdprofilewatcher.go